src/include/miopen/nllloss/problem_description.hpp

```diff
@@ -0,0 +1,261 @@
+            std::transform(strides.begin(),
+                           strides.end(),
+                           lengths.begin(),
+                           std::back_inserter(p),
+                           [](size_t a, size_t b) { return std::make_pair(a, b); });
+            std::sort(p.begin(), p.end());
+            for(int i = 1; i < p.size(); ++i)
+            {
+                if(p[i].first != p[i - 1].first * p[i - 1].second)
+#if MIOPEN_BUILD_DEV || !MIOPEN_NDEBUG
+                    MIOPEN_THROW(miopenStatusBadParm, "NLLLoss: Tensor strides do not match.");
+#else
+                    return false;
+#endif
+            }
+            return true;
+        };
+        return isRightStride(inputDesc) && isRightStride(targetDesc) && isRightStride(outputDesc) &&
+               isRightStride(weightDesc);
+    }
+
+    bool IsAllContiguous() const
+    {
+        auto isContiguous = [](TensorDescriptor td) {
+            size_t s = 1;
+            for(int i = td.GetSize() - 1; i >= 0; --i)
+            {
+                if(s != td.GetStrides()[i])
+                {
+                    return false;
+                }
+                s *= td.GetLengths()[i];
+            }
+            return true;
+        };
+        return isContiguous(inputDesc) && isContiguous(targetDesc) && isContiguous(weightDesc) &&
+               isContiguous(outputDesc);
+    }
+
+    bool IsAllPacked() const
+    {
+        if(!(inputDesc.IsPacked() && targetDesc.IsPacked() && weightDesc.IsPacked() &&
+             outputDesc.IsPacked()))
+        {
+#if MIOPEN_BUILD_DEV || !MIOPEN_NDEBUG
+            MIOPEN_THROW(miopenStatusBadParm, "NLLLoss: Unpacked tensors not supported.");
+#else
+            return false;
+#endif
+        }
+        return true;
+    }
+
+protected:
+    TensorDescriptor inputDesc;
+    TensorDescriptor targetDesc;
+    TensorDescriptor weightDesc;
+    TensorDescriptor outputDesc;
+
+    int32_t ignore_index;
+    bool is_fwd;
+
+    NetworkConfig MakeForwardNetworkConfig() const;
+};
+
+struct UnreduceProblemDescription : ProblemDescription
+{
+    UnreduceProblemDescription(const TensorDescriptor& inputDesc_,
+                               const TensorDescriptor& targetDesc_,
+                               const TensorDescriptor& weightDesc_,
+                               const TensorDescriptor& outputDesc_,
+                               int32_t ignore_index_,
+                               bool is_fwd_)
+        : ProblemDescription(
+              inputDesc_, targetDesc_, weightDesc_, outputDesc_, ignore_index_, is_fwd_)
+    {
+        IsSameType();
+        IsRightLength();
+        IsRightStride();
+        IsAllPacked();
+    }
+
+    size_t GetNtotal() const { return outputDesc.GetElementSize(); }
+    size_t GetC() const { return weightDesc.GetElementSize(); }
+
+    bool IsSameType() const
+    {
+        if(inputDesc.GetType() != weightDesc.GetType())
+        {
+#if MIOPEN_BUILD_DEV || !MIOPEN_NDEBUG
```

Comment:
remove this part refer https://github.com/ROCm/MIOpen/pull/2766#discussion_r1575181661